File: scripts/gen-icons.py
#!/usr/bin/env python3
"""Generate extension icons as <#> — HTML angle brackets in teal, Markdown # in orange."""

from PIL import Image, ImageDraw, ImageFont
from pathlib import Path

SIZES = [16, 48, 128]
OUT_DIR = Path(__file__).parent.parent / 'icons'

FONT_PATHS = [
    '/System/Library/Fonts/Menlo.ttc',
    '/System/Library/Fonts/Courier New.ttf',
    '/Library/Fonts/Courier New.ttf',
]

BG     = (20, 20, 32, 255)
TEAL   = (77, 208, 225)   # < >  HTML
ORANGE = (255, 153, 0)    # #    Markdown

def load_font(size: int) -> ImageFont.FreeTypeFont:
    for path in FONT_PATHS:
        try:
            return ImageFont.truetype(path, size)
        except OSError:
            continue
    return ImageFont.load_default()

def make_icon(size: int) -> Image.Image:
    img  = Image.new('RGBA', (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)

    radius = max(2, size // 7)
    draw.rounded_rectangle([(0, 0), (size - 1, size - 1)], radius=radius, fill=BG)

    font    = load_font(int(size * 0.48))
    parts   = [('<', TEAL), ('#', ORANGE), ('>', TEAL)]
    total_w = sum(draw.textlength(ch, font=font) for ch, _ in parts)
    bbox    = draw.textbbox((0, 0), '<#>', font=font)
    text_h  = bbox[3] - bbox[1]

    x = (size - total_w) / 2
    y = (size - text_h) / 2 - bbox[1]

    for ch, color in parts:
        draw.text((x, y), ch, font=font, fill=color)
        x += draw.textlength(ch, font=font)

    return img

if __name__ == '__main__':
    OUT_DIR.mkdir(exist_ok=True)
    for size in SIZES:
        path = OUT_DIR / f'icon{size}.png'
        make_icon(size).save(path)
        print(f'  {path}')

File: scripts/gen-icons.ts
import sharp from 'sharp'
import { mkdirSync } from 'fs'
import { resolve, dirname } from 'path'
import { fileURLToPath } from 'url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const OUT_DIR = resolve(__dirname, '../icons')

const SIZES = [16, 48, 128]
const TEAL   = '#4DD0E1'
const ORANGE = '#FF9900'

function makeSvg(size: number): Buffer {
  const fs = Math.round(size * 0.72)
  const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}">` +
    `<text x="${size / 2}" y="${size / 2}" ` +
    `font-family="Menlo,'Courier New',monospace" font-size="${fs}" ` +
    `text-anchor="middle" dominant-baseline="central">` +
    `<tspan fill="${TEAL}">&lt;</tspan>` +
    `<tspan fill="${ORANGE}">#</tspan>` +
    `</text></svg>`
  return Buffer.from(svg)
}

async function main() {
  mkdirSync(OUT_DIR, { recursive: true })
  for (const size of SIZES) {
    const out = resolve(OUT_DIR, `icon${size}.png`)
    await sharp(makeSvg(size)).png().toFile(out)
    console.log(`  ${out}`)
  }
}

main()
